crates/model/src/defi/block.rs

```diff
@@ -0,0 +1,372 @@
+// -------------------------------------------------------------------------------------------------
+//  Copyright (C) 2015-2025 Nautech Systems Pty Ltd. All rights reserved.
+//  https://nautechsystems.io
+//
+//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
+//  You may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+// -------------------------------------------------------------------------------------------------
+
+use std::fmt::{Display, Formatter};
+
+use nautilus_core::{UnixNanos, datetime::NANOSECONDS_IN_SECOND};
+use serde::{Deserialize, Deserializer};
+use ustr::Ustr;
+
+/// Represent Ethereum-compatible compatible blockchain block with essential metadata
+#[derive(Debug, Clone, Deserialize)]
+pub struct Block {
+    /// The unique identifier hash of the block
+    hash: Ustr,
```

Comment:
Ustr is best used for values that have a limited number of unique values e.g. instrument id. BTC has about 1M blocks, others probably more which means 1M unique global allocations that will remain alive for the duration of the program. It means that even if block information is streamed and aggregated the blocks unique hash will remain in the global memory.

Reply:
Thanks @twitu I'll tweak this after merge.